### PR TITLE
Oversampling Mitigation

### PR DESCRIPTION
**Pull request by atshaw43**
*Issue #, if available:*

*Description of changes:*
Adding Oversampling Mitigation.
SQSMessageHelper's IsSampled function returns true only if the SQSMessage passed in is sampled.
BeginSubsegmentWithoutSampling was added to AWSXrayReccorder which allows the creating of subsegments that ignore the parent entity's sampled flag and never emit.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by srprash** [comment]
Also, I would really appreciate if you can add description about the issue being addressed and the changes made in this PR. I may have the context now but will be helpful if we ever revisit it in the future. 😄

File: sdk/src/Core/AWSXRayRecorder.netcore.cs [hunk 1]
```diff
@@ -405,7 +405,7 @@ private void EndFacadeSegment()
                 if (!IsTracingDisabled())
                 {
                     PrepEndSegment(facadeSegment);
-                    if (facadeSegment.Sampled == SampleDecision.Sampled && facadeSegment.RootSegment != null && facadeSegment.RootSegment.Size >= 0)
```
[the remark this answers is not in the document]
Comment (atshaw43) [reply sdk/src/Core/AWSXRayRecorder.netcore.cs:408]: I moved where the check is. It is now in the stream itself and based on the subsegment. Which is based on the facadeSegment here unless we override it.

File: sdk/src/Core/AWSXRayRecorder.Core.csproj [hunk 2]
```diff
@@ -1,7 +1,7 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>net45;net462;netstandard2.0;netcoreapp3.1;net6.0</TargetFrameworks>
+    <TargetFrameworks>net6.0;net45;net462;netstandard2.0;netcoreapp3.1</TargetFrameworks>
```
[the remark this answers is not in the document]
Comment (atshaw43) [reply sdk/src/Core/AWSXRayRecorder.Core.csproj:4]: Done

File: sdk/src/Core/AWSXRayRecorderImpl.cs [hunk 3]
```diff
@@ -240,6 +240,16 @@ public void EndSegment(DateTime? timestamp = null)
         /// <exception cref="EntityNotAvailableException">Entity is not available in trace context.</exception>
         public abstract void BeginSubsegment(string name, DateTime? timestamp = null);
 
+        /// <summary>
+        /// Begin a tracing subsegment. A new segment will be created and added as a subsegment to previous segment.
```
[the remark this answers is not in the document]
Comment (atshaw43) [reply sdk/src/Core/AWSXRayRecorderImpl.cs:244]: Thanks, updated the original BeginSubsegment as well!

File: sdk/test/UnitTests/TestLambdaContext.cs [hunk 4]
```diff
@@ -48,6 +55,53 @@ public new void TestCleanup()
...
+
+                LambdaTestHelper(recorder, true);
+                mockEmitter.Verify(x => x.Send(It.IsAny<Subsegment>()), Times.Exactly(4));
+
+                LambdaTestHelper(recorder, false);
+                mockEmitter.Verify(x => x.Send(It.IsAny<Subsegment>()), Times.Exactly(4));
+            }
+        }
+
+        private static void LambdaTestHelper(AWSXRayRecorder recorder, bool sampled)
```
Comment (carolabadeer) [inline sdk/test/UnitTests/TestLambdaContext.cs:81]: I think this test should be renamed to follow the naming pattern of the other tests. e.g: `TestLambdaEnv` or `TestLambdaHelper`

File: sdk/test/UnitTests/TestLambdaContext.cs [hunk 5]
```diff
@@ -22,12 +22,19 @@
...
+using Moq;
+using Amazon.XRay.Recorder.Core.Internal.Emitters;
+using Amazon.XRay.Recorder.UnitTests.Tools;
 
 namespace Amazon.XRay.Recorder.UnitTests
 {
     [TestClass]
     public class TestLambdaContext : TestBase
     {
+        private const string URL = "https://httpbin.org/";
```
Comment (carolabadeer) [inline sdk/test/UnitTests/TestLambdaContext.cs:37]: Rename to Mock URL to clarify that this URL is used to mock an HTTP request. 
Is there a specific reason this was set as a class variable and not a local variable in the test it is being used in?

File: sdk/src/Core/AWSXRayRecorderImpl.cs [hunk 6]
```diff
@@ -240,6 +240,16 @@ public void EndSegment(DateTime? timestamp = null)
...
+        /// </summary>
+        /// <param name="name">Name of the operation.</param>
+        public void BeginSubsegmentWithoutSampling(string name)
+        {
+            BeginSubsegment(name, null);
```
[the remark this answers is not in the document]
Comment (atshaw43) [reply sdk/src/Core/AWSXRayRecorderImpl.cs:249]: Done

File: sdk/test/UnitTests/TestLambdaContext.cs (hunk 4 above)
Comment (atshaw43) [reply sdk/test/UnitTests/TestLambdaContext.cs]: Done

File: sdk/test/UnitTests/TestLambdaContext.cs (hunk 5 above)
Comment (atshaw43) [reply sdk/test/UnitTests/TestLambdaContext.cs]: Done

File: sdk/test/UnitTests/AWSXRayRecorder.UnitTests.csproj [hunk 7]
```diff
@@ -2,7 +2,7 @@
 
   
   <PropertyGroup>
-    <TargetFrameworks>net452;net462;netcoreapp2.0;netcoreapp3.1;net6.0</TargetFrameworks>
+    <TargetFrameworks>net6.0;net452;net462;netcoreapp2.0;netcoreapp3.1</TargetFrameworks>
```
[the remark this answers is not in the document]
Comment (atshaw43) [reply sdk/test/UnitTests/AWSXRayRecorder.UnitTests.csproj]: Done

File: sdk/src/Core/Lambda/SQSMessageHelper.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Net;
+using System.Net.Http;
+using System.Text;
+using Amazon.Lambda.SQSEvents;
+using Amazon.Runtime.Internal.Util;
+using ThirdParty.LitJson;
```
[the remark this answers is not in the document]
Comment (atshaw43) [reply sdk/src/Core/Lambda/SQSMessageHelper.cs]: Done

File: sdk/src/Core/IAWSXRayRecorder.cs [hunk 9]
```diff
@@ -95,6 +95,13 @@ public interface IAWSXRayRecorder : IDisposable
         /// <param name="timestamp">Sets the start time for the subsegment</param>
         void BeginSubsegment(string name, DateTime? timestamp = null);
 
+        /// <summary>
+        /// Start a subsegment with a given segment
+        /// This subsegment will not emit and its trace context will have Sampled=0
+        /// </summary>
+        /// <param name="name">Name of the operation</param>
```
[the remark this answers is not in the document]
Comment (atshaw43) [reply sdk/src/Core/IAWSXRayRecorder.cs]: Done

File: sdk/src/Core/IAWSXRayRecorder.cs [hunk 10]
```diff
@@ -95,6 +95,13 @@ public interface IAWSXRayRecorder : IDisposable
         /// <param name="timestamp">Sets the start time for the subsegment</param>
         void BeginSubsegment(string name, DateTime? timestamp = null);
 
+        /// <summary>
+        /// Start a subsegment with a given segment
```
[the remark this answers is not in the document]
Comment (atshaw43) [reply sdk/src/Core/IAWSXRayRecorder.cs]: Done, and fixed the original BeginSegment.

**Review by srprash — APPROVED** [review]
LGTM.

**Comment by atshaw43** [comment]
Example Code:

```
using Amazon.Lambda.Core;
using Amazon.Lambda.SQSEvents;
using Amazon.XRay.Recorder.Core;
using Amazon.XRay.Recorder.Core.Lambda;

[assembly: LambdaSerializer(typeof(Amazon.Lambda.Serialization.Json.JsonSerializer))]

namespace MyFunction;

public class Function
{
    public string HandleSQSEvent(SQSEvent sqsEvent, ILambdaContext context)
    {
        bool isSampled = false;

        foreach (SQSEvent.SQSMessage sqsMessage in sqsEvent.Records)
        {
            if (SQSMessageHelper.IsSampled(sqsMessage))
            {
                isSampled = true;
            }
        }

        if (isSampled)
        {
            AWSXRayRecorder.Instance.BeginSubsegment("Test 1");
        }
        else
        {
            AWSXRayRecorder.Instance.BeginSubsegmentWithoutSampling("Test 2");
        }

        // Do my processing work here
        Console.WriteLine("Doing work");

        // End my subsegment
        AWSXRayRecorder.Instance.EndSubsegment();

        return "Done";
    }
}
```

**Pull request merged**